dimagi/utils/couch/bulk.py

```diff
@@ -100,3 +100,33 @@ def wrapped_docs(cls, keys):
     rows = get_docs(cls.get_db(), keys)
     for row in rows:
         yield cls.wrap(row['doc'])
+
+
+class IterativeSaver(object):
+    """
+    Bulk save doc dicts in chunks.  Works well with the output from iter_docs.
+
+        with IterativeSaver(db) as iter_db:
+            for doc in iter_docs(db)
+                iter_db.save(doc)
+    """
+    def __init__(self, database, chunksize=100):
+        self.db = database
+        self.chunksize = chunksize
+
+    def __enter__(self):
+        self.to_save = []
+        return self
+
+    def commit(self):
+        self.db.bulk_save(self.to_save)
```

Comment:
This will raise `BulkSaveError` if there's a resource conflict (even though all the docs that did _not_ confict will have gone through). If you catch the `BulkSaveError` as `e`, you can use `e.errors` to get the ids of all the docs that did not get saved. I would keep a running list of these and then raise your own `BulkSaveError` with those or have some other way for the caller to access that list.